third_party/binutils/test_decenc/addr16.asm

```diff
@@ -0,0 +1,22 @@
+
+#undef FUNCNAME
+#define FUNCNAME test_addr16_s_asm
+DECLARE_FUNC_SEH(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        /* There is intentionally no prologue and epilogue. This function is not
+         * supposed to be actually called. We are using the function like a label
+         * for the decoder.
+         */
+        END_PROLOG
+        RAW(67) RAW(a0) RAW(98) RAW(08) RAW(67) RAW(66)
```

Comment:
So I went and looked at my (old I guess, 2.18.50) binutils sources I happen to have sitting around, and there I see:
```
 $ cat gas/testsuite/gas/i386/addr16.s
	.text
	addr16 mov	0x0898,%al
	addr16 mov	0x0898,%ax
	addr16 mov	0x0898,%eax
	addr16 mov	%al,0x0898
	addr16 mov	%ax,0x0898
	addr16 mov	%eax,0x0898
```

That is readable.  Why were those converted into raw bytes which no human can read?  This does not seem conducive to code reading or understanding, to updating this file if binutils augments or changes their version, etc.

Reply:
I understand why you might convert to raw bytes for avx-512 or other recent ISA features where a test machine may not be able to assemble if in assembler form, but not for this test.

An alternative to assembling at test time would be to take the *.s files and assemble them yourself once into a binary using a checked-in script that is run with a recent assembler, so we don't need *any* raw entries and can use the binutils .s files as-is.  The test then decodes from a public symbol in a checked-in binary.  Maybe it's too complex and we'd only do this if we expected to updated from binutils frequently and really wanted to have unchanged .s files from them.

Reply:
The raw bytes come from the .d results files in the binutils testsuite. The binutils suite assembles first, we don't have an assembler that reads at&t or any other syntax.

I don't see that we ever have to change these files even if binutils changes. It's all valid x86 encodings.

Reply:
> we don't have an assembler that reads at&t or any other syntax.

?  Sure we do: we have a number of assembly files which we have to assemble as part of building DR and at least on Linux AT&T is perfectly supported by the toolchain we rely on.

Reply:
You mean if we use the local toolchain. Sure. That's another way how to do it. I (wanted to) made (make) it independent from using an assembler.

Reply:
> You mean if we use the local toolchain. Sure. That's another way how to do it. I (wanted to) made (make) it independent from using an assembler.

But it is already using an assembler, and the local toolchain: the local assembler is taking these RAW declarations and creating binary code.  Is the reason for using RAW to make it build on Windows where it was easier to use RAW than to convert the AT&T into Intel?

Reply:
As a side note, I think we're testing our decoder/encoder one way or the other.

Windows was not the reason. But that's a good reason, too. I think the gas test suite is testing something different than us. They are testing the gas assembler, i.e. taking mnemonics and corner cases of mnemonics and encoding them. Then they cross check the results with objdump. For DynamoRIO, we're testing the raw-bytes decoder (and the encoder kind of by making a round-trip). So I looked at the binutils test suite just as a source to get interesting x86 encoding bytes.

The fact that they're testing gas includes invalid cases. If you took the .s files, you had to filter out the _invalid_ cases first. That's in addition to the _valid_ cases we do not support (e.g. CPL-0 instructions). I already had to filter the encodings for all the cases we either don't support or not yet support (i#3577, i#3578, i#1312, i#3581). I didn't want to filter the mnemonics as well, that don't add much except, yes, readability.

Reply:
The DR decoder does support CPL-0 instructions.

Reply:
Readability is a big deal.   Making these just a sequence of hex numbers makes them unreadable and disconnects them from the original files.  If they're just piles of hex numbers disconnected from the binutils files I would much prefer to have them in one file instead of dealing with 400 (!) separate tiny unreadable files.

Reply:
Not sure what you mean by "just piles of hex numbers". They could be "just piles of mnemonics" then as well. How do you want to handle the currently disabled subtests then? Right now they are conveniently listed in suite/tests/CMakeLists.txt. If I pipe them into one big file with comments so that I can see which ones were which, do you want to ifdef out the parts that are currently disabled per DISABLED_UNTIL_BUG_<N>?

Reply:
> The DR decoder does support CPL-0 instructions.

Ok, I did see some that the decoder does not have, I assumed we didn't generally support ring 0 instructions. I will add them to DDK-3578 (for example clac, slac).

Reply:
@Carrotman42 any opinion on having these 400 separated raw-hex files each as a separate test using ifdefs in one combined file, vs inlining into that file for a single test?

Reply:
I don't have a problem with converting it to a one file setup (just with re-writing everything using .s files). Done.